rust_src/src/eval.rs

```diff
@@ -1338,4 +1338,135 @@ pub fn condition_case(args: LispCons) -> LispObject {
+    unsafe { pdl.let_.as_ref().old_value }
+}
+
+#[no_mangle]
+pub extern "C" fn set_specpdl_old_value(mut pdl: SpecbindingRef, val: LispObject) {
+    debug_assert!(pdl.kind() >= SPECPDL_LET);
+    unsafe {
+        pdl.let_.as_mut().old_value = val;
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn default_toplevel_binding(symbol: LispObject) -> SpecbindingRef {
+    let current_thread = ThreadState::current_thread();
+    let specpdl = SpecbindingRef::new(current_thread.m_specpdl);
+
+    let mut binding = SpecbindingRef::new(std::ptr::null_mut());
+    let mut pdl = SpecbindingRef::new(current_thread.m_specpdl_ptr);
+
+    while pdl > specpdl {
+        unsafe {
+            pdl.ptr_sub(1);
+        }
+        match pdl.kind() {
+            SPECPDL_LET_DEFAULT | SPECPDL_LET => {
+                if specpdl_symbol(pdl) == symbol {
+                    binding = pdl.clone()
+                }
+            }
+            SPECPDL_UNWIND | SPECPDL_UNWIND_PTR | SPECPDL_UNWIND_INT | SPECPDL_UNWIND_VOID
+            | SPECPDL_BACKTRACE | SPECPDL_LET_LOCAL => {}
+            _ => panic!("Incorrect specpdl kind"),
+        }
+    }
+
+    binding
+}
+
+/// Define SYMBOL as a variable, and return SYMBOL.
+/// You are not required to define a variable in order to use it, but
+/// defining it lets you supply an initial value and documentation, which
+/// can be referred to by the Emacs help facilities and other programming
+/// tools.  The `defvar' form also declares the variable as \"special\",
+/// so that it is always dynamically bound even if `lexical-binding' is t.
+///
+/// If SYMBOL's value is void and the optional argument INITVALUE is
+/// provided, INITVALUE is evaluated and the result used to set SYMBOL's
+/// value.  If SYMBOL is buffer-local, its default value is what is set;
+/// buffer-local values are not affected.  If INITVALUE is missing,
+/// SYMBOL's value is not set.
+///
+/// If SYMBOL has a local binding, then this form affects the local
+/// binding.  This is usually not what you want.  Thus, if you need to
+/// load a file defining variables, with this form or with `defconst' or
+/// `defcustom', you should always load that file _outside_ any bindings
+/// for these variables.  (`defconst' and `defcustom' behave similarly in
+/// this respect.)
+///
+/// The optional argument DOCSTRING is a documentation string for the
+/// variable.
+///
+/// To define a user option, use `defcustom' instead of `defvar'.
+/// usage: (defvar SYMBOL &optional INITVALUE DOCSTRING)
+#[lisp_fn(min = "1", unevalled = "true")]
+pub fn defvar(args: LispCons) -> LispObject {
+    let (sym_obj, tail) = args.into();
+
+    if tail.is_not_nil() {
+        if LispCons::from(tail).length() > 2 {
+            error!("Too many arguments");
+        }
+
+        let sym: LispSymbolRef = sym_obj.into();
+        let tem = default_boundp(sym);
+
+        /* Do it before evaluating the initial value, for self-references.  */
```

Comment:
Comments should be with two slashes, as in `// Do it before...`

Reply:
Done